### PR TITLE
Clamwin gui cpp

### PR DESCRIPTION
**Pull request by alexcherney**
This pull request introduces a new version (1.5.2) of ClamAV Win32, with a major focus on adding a native Win32 C++ GUI frontend and its dedicated test suite. It also includes improvements to build configuration, documentation, and licensing. The most important changes are grouped below.

### GUI Integration and Testing

* Added `cmake/clamwin_gui.cmake` to build the new native Win32 C++ GUI (`clamwin`) and its test executable (`clamwin_gui_test.exe`), including test targets for real tool smoke tests and stricter update probes.
* Introduced a build version increment script (`cmake/increment_clamwin_build_version.cmake`) to generate version headers and RC files for the GUI, ensuring consistent build numbering.
* Updated `CMakeLists.txt` to include the GUI build logic and dependencies, and set up source file properties for helper modules.
* Added detailed GUI test workflow instructions and build/test commands to `README.md` and a new `build-clamav-win32.md` file, covering configuration, smoke tests, and opt-in negative/stricter probes. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R106-R145) [[2]](diffhunk://#diff-1cd50e5f70c4b0a21c2115b89d4f172b6ba64188bbefdb686decc6f605fa71d2R1-R170)

### Versioning and Licensing

* Bumped project version to 1.5.2 in `CMakeLists.txt` and `README.md`. [[1]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL2-R2) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L1-R1)
* Added the MIT license for the third-party `doctest` library used in GUI tests.

### Build Configuration and Compatibility

* Improved Windows version targeting in `cmake/curl.cmake` by forcing the minimum version string for older Windows builds.
* Enhanced Rust static library filtering in `cmake/legacy.cmake` with flexible Python interpreter detection, improving cross-platform compatibility.
* Made zlib-ng usage on x86 builds configurable via a CMake option in `cmake/zlib.cmake`.
* Allowed explicit cert path configuration for FreshClam in `README.md`, supporting portable/non-default layouts. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R88-R89) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R76)

---

**References:** [[1]](diffhunk://#diff-6acdb537552e3a561d6fe1d283e111d8dfc6029715f124b834ce5c16c2ae073eR1-R183) [[2]](diffhunk://#diff-1fa44ea300c450d2c42347496b621e6b14ee68f6ddbae10a975fdc66cb9edda1R1-R66) [[3]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR157-R158) [[4]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R106-R145) [[5]](diffhunk://#diff-1cd50e5f70c4b0a21c2115b89d4f172b6ba64188bbefdb686decc6f605fa71d2R1-R170) [[6]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL2-R2) [[7]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L1-R1) [[8]](diffhunk://#diff-c3cd28fd32a0d5c70a60a1a91f0e1257d2a6e2df8c57baecbd94be03ef01c5dcR1-R21) [[9]](diffhunk://#diff-cf2e74f2dad2eb59515b6f6cee50e89c151e5fb018948f763a3e1b367dea519dL49-R49) [[10]](diffhunk://#diff-b1b859e52c1d1ad436ff2e832f5f3730d724c1081a647dcf4f27315aaa50a1a5L90-R110) [[11]](diffhunk://#diff-9c0008f51a59d8b1a69304775d8463a8a77a101b3db3e5ad9e77f1f3ba2ae2f6L1-R3) [[12]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R88-R89) [[13]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R76)

### CODE REVIEW

**Comment by alexcherney** [comment]
@sherpya here's the built setup of the alpha version
https://github.com/clamwin/update-test/blob/alpha/clamwin-1.5.2-alpha-7-setup-nodb.exe

**Comment by sherpya** [comment]
> @sherpya here's the built setup of the alpha version https://github.com/clamwin/update-test/blob/alpha/clamwin-1.5.2-alpha-7-setup-nodb.exe

are you happy with copilot bot? lol :)

File: src/clamwin-gui-cpp/cw_update_checker.cpp [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+void CWUpdateChecker::waitForThread()
+{
+    if (m_hThread)
+    {
+        WaitForSingleObject(m_hThread, 15000);
+        CloseHandle(m_hThread);
+        m_hThread = NULL;
+    }
+}
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_update_checker.cpp:108]: Addressed. `waitForThread()` now waits for thread completion before closing the handle (`WaitForSingleObject(..., INFINITE)`), preventing timeout-driven teardown while the worker can still be active.

File: src/clamwin-gui-cpp/cw_process.cpp [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    {
+        WaitForSingleObject(m_hThread, 5000);
+        CloseHandle(m_hThread);
+        m_hThread = NULL;
+    }
+
+    /* Reader thread closes handles; just free ctx */
+    delete m_ctx;
+    m_ctx = NULL;
+}
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_process.cpp:216]: Fixed. `CWProcess::stop()` now joins the reader thread with `INFINITE` wait before context teardown, so `m_ctx` is no longer freed while the reader thread may still access it.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+	// duplicate the object pointer and registry handle
+	if (pDataObj)
+	{
+		m_pDataObj = pDataObj;
+		pDataObj->AddRef();
+	}
+
+	// use the given IDataObject to get a list of filenames (CF_HDROP)
+	hres = pDataObj->GetData(&fmte, &medium);
+
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp:86]: Good catch, fixed. `Initialize(...)` now returns `E_INVALIDARG` when `pDataObj` is null before any dereference (`GetData` call), preventing null-pointer crashes.

File: src/clamwin-gui-cpp/cwdefs.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#ifndef CWDEFS_H
+#define CWDEFS_H
+
+/* Target Windows XP (5.1) minimum — for InitCommonControlsEx, etc.
+#ifndef _WIN32_WINNT
+#  define _WIN32_WINNT 0x0410
+#endif
+
+#ifndef WINVER
+#  define WINVER 0x0410
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cwdefs.h]: Fixed in latest changes. I closed the unterminated comment and aligned the macros with the XP (5.1) target:
- `_WIN32_WINNT` set to `0x0501`
- `WINVER` set to `0x0501`

This removes the accidental comment-out and keeps the platform defines consistent.

File: src/clamwin-gui-cpp/cw_update_checker.cpp [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+/* ─── Hardcoded endpoints ───────────────────────────────────── */
+
+static const char* CW_GITHUB_API_HOST = "api.github.com";
+static const char* CW_GITHUB_API_PATH = "/repos/clamwin/update-test/releases/latest";
+static const char* CW_DOWNLOAD_PAGE   = "https://www.clamwin.com/download";
+
+const char* CWUpdateChecker::apiUrl()
+{
+    return "https://api.github.com/repos/clamwin/clamwin/releases/latest";
+}
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_update_checker.cpp:44]: Fixed. I aligned the runtime endpoint constants with `apiUrl()` so both point to `clamwin/clamwin` releases:
- `CW_GITHUB_API_PATH` now uses `/repos/clamwin/clamwin/releases/latest`

This removes the repo mismatch between the declared API URL and actual request path.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+        FALSE,            // Set handle inheritance to FALSE.
+        0,                // No creation flags.
+        NULL,             // Use parent's environment block.
+        NULL,             // Use parent's starting directory.
+        &si,              // Pointer to STARTUPINFO structure.
+        &pi )             // Pointer to PROCESS_INFORMATION structure.
+    )
+    {
+        TCHAR szMsg[MAX_PATH*2];
+        _sntprintf(szMsg, sizeof(szMsg), _T("Error: Unable to execute command %s."), command.c_str());
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp]: Fixed. Updated `_sntprintf` size argument from `sizeof(szMsg)` to `_countof(szMsg)` so it uses TCHAR element count (correct in Unicode builds).

File: src/clamwin-gui-cpp/tests/test_scheduler.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        time_t past = time(NULL) - (3 * 86400);
+        cfg.scanLastRunTime = past;
+        cfg.updateLastRunTime = past;
+
+        // Create a dummy message loop window to catch postmessages
+        HWND hwnd = CreateWindowA("STATIC", "Test", 0, 0, 0, 0, 0, NULL, NULL, GetModuleHandle(NULL), NULL);
+        REQUIRE(hwnd != NULL);
+
+        long long dummy_scan_last = past;
+        time_t now = time(NULL);
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/tests/test_scheduler.cpp]: Fixed. The integration test now uses a fixed mocked timestamp via `mktime(...)` and aligns schedule fields (`scanHour/scanMinute/scanDay` and update counterparts) to that mock time. It no longer depends on `time(NULL)`.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+    }
+	// allocate memory for our combined path
+	// add length of [ --path=""]
+	INT cbPath = (MAX_PATH + 10) * numFiles;
+	m_szPath = new TCHAR[cbPath];
+	m_szPath[0] = _T('\0');
+	for(int i = 0; i < numFiles; i++)
+	{
+		DragQueryFile((HDROP)medium.hGlobal, i, szPath, sizeof(szPath));
+		// check the long path validity for a unicode build
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp:113]: Fixed. I changed TCHAR buffer-size arguments to character-count form:
- `DragQueryFile(..., _countof(szPath))`
- `GetShortPathNameW(..., _countof(wtemp))`

This keeps behavior correct for Unicode builds.

File: src/clamwin-gui-cpp/cw_gui_shared.h [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+#include <string.h>
+
+#include "cw_build_version.h"
+
+#ifndef CLAMWIN_VERSION_STR
+#define CLAMWIN_VERSION_STR     "1.5.2.0"
+#endif
+#define CLAMWIN_APPNAME         "ClamWin Free Antivirus"
+#define CLAMWIN_APPNAME_W       L"ClamWin Free Antivirus"
+#define CLAMWIN_WEBSITE         "http://www.clamwin.com"
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_gui_shared.h]: Updated. `CLAMWIN_WEBSITE` now uses HTTPS (`https://www.clamwin.com`) for consistency with the rest of update URL policy.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+#endif
+		// convert \ to / so cygwin doesn't go crazy (particularly over UNC names)
+//		_tcsreplace(szPath, _T('\\'), _T('/'));
+		len = _tcslen(szPath);
+		// remove last slash from the scanning path
+        if(szPath[len-1] == _T('\\'))
+            szPath[len-1] = _T('\0');
+        _tcsncat(m_szPath, _T(" --path=\""), cbPath);
+        _tcsncat(m_szPath, szPath, cbPath);
+        _tcsncat(m_szPath, _T("\""), cbPath);
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp]: Addressed. I removed the unsafe `_tcsncat` sequence and switched to bounded append formatting while tracking used/capacity, with explicit NUL termination. This avoids destination overflow when many/long paths are selected.

**Comment by alexcherney** [comment]
> > @sherpya here's the built setup of the alpha version https://github.com/clamwin/update-test/blob/alpha/clamwin-1.5.2-alpha-7-setup-nodb.exe
> 
> are you happy with copilot bot? lol :)

actually - yes. With right supervision it saves heaps of time :)
are you?

**Comment by alexcherney** [comment]
> actually - yes. With right supervision it saves heaps of time :) are you?
@sherpya my main pain point was that keeping the toolset to build clamwin was next to impossible. Now the only legacy tool is Innosetup 5.23, the rest builds CMake and mingw on modern hardware.
Still needs heaps of testing but is close to being usable :)
And should run on XP and Win98

**Comment by sherpya** [comment]
> > actually - yes. With right supervision it saves heaps of time :) are you?
> > @sherpya my main pain point was that keeping the toolset to build clamwin was next to impossible. Now the only legacy tool is Innosetup 5.23, the rest builds CMake and mingw on modern hardware.
> > Still needs heaps of testing but is close to being usable :)
> > And should run on XP and Win98

win98 has a lot of problem, it can run only 1.4.x, 32bit versions anyway have very bad time at loading the big database.
I suppose we are unable to provide support at least for win98, I think you can move completely to unicode functions, btw win98 legacy version (1.4.x) bakes in a modified version of opencow.

**Comment by alexcherney** [comment]
> > > actually - yes. With right supervision it saves heaps of time :) are you?
> > > @sherpya my main pain point was that keeping the toolset to build clamwin was next to impossible. Now the only legacy tool is Innosetup 5.23, the rest builds CMake and mingw on modern hardware.
> > > Still needs heaps of testing but is close to being usable :)
> > > And should run on XP and Win98
> 
> win98 has a lot of problem, it can run only 1.4.x, 32bit versions anyway have very bad time at loading the big database. I suppose we are unable to provide support at least for win98, I think you can move completely to unicode functions, btw win98 legacy version (1.4.x) bakes in a modified version of opencow.

yeah, lots of hard work to make it run but then there is nowhere near enough memory to hold the db
do we say WinXP as the minimum supported version?

File: src/clamwin-gui-cpp/.agent/workflows/testing.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
+---
+description: How to run tests and add new tests for the ClamWin GUI
+---
+
+# Testing Workflow
+
+## Strategy: MinUnit + Console Test Runner
+We use a minimal C unit test approach (MinUnit macros in `minunit.h`) with a console test runner.
+The test executable (`clamwin_test.exe`) is a separate build target that links the same source files
+as the GUI but replaces `clamwin_main.c` with `test_main.c`.
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/.agent/workflows/testing.md]: Updated. Rewrote the workflow doc to match the current doctest-based setup (`clamwin_gui_test`, `clamwin_gui_check`, `clamwin_gui_check_real_tools`) and removed stale MinUnit guidance.

File: src/clamwin-gui-cpp/cw_scheduler.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    if (ran_this_interval) return 0;
+
+    /* Is today an eligible day? */
+    bool eligible_day = false;
+    switch (frequency) {
+        case 0: /* Daily */     eligible_day = true; break;
+        case 1: /* Workdays */  eligible_day = (tm->tm_wday >= 1 && tm->tm_wday <= 5); break;
+        case 2: /* Weekly */    eligible_day = (tm->tm_wday == day); break;
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_scheduler.cpp]: Fixed. Weekly scheduling now converts config day numbering (`0=Mon..6=Sun`) to `tm_wday` (`0=Sun..6=Sat`) before comparison, eliminating the weekday mapping mismatch.

File: src/clamwin-gui-cpp/shell-extension/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    ${CMAKE_CURRENT_SOURCE_DIR}
+    ${CLAMWIN_DIR}/src/clamwin-gui-cpp
+)
+
+target_compile_definitions(clamwin_shell_extension PRIVATE
+    ${CLAMWIN_DEFINES}
+    _CRT_SECURE_NO_WARNINGS
+)
+
+if(CLAMWIN_SHELLEXT_UNICODE)
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/CMakeLists.txt]: Fixed. `CLAMWIN_SHELLEXT_UNICODE` is now compared explicitly as a string (`if(CLAMWIN_SHELLEXT_UNICODE STREQUAL "ON")`), so `OFF` no longer evaluates truthy and Win98 ANSI builds stay ANSI.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension.h [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+//
+// class id:  65713842-C410-4f44-8383-BFE01A398C90
+//
+//
+
+#ifndef CW_SHELL_EXTENSION_H
+#define CW_SHELL_EXTENSION_H
+
+#define CF_HDROP        15
+
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension.h]: Fixed. Removed the local `#define CF_HDROP 15` and now rely on the platform header definition to avoid redefinition issues.

File: src/clamwin-gui-cpp/cw_scheduler.cpp [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    /* Determine current interval identifier to prevent running twice */
+    bool ran_this_interval = false;
+    if (last_rt > 0) {
+        if (frequency == 3) {
+            ran_this_interval = (ltm.tm_year == tm->tm_year && ltm.tm_yday == tm->tm_yday && ltm.tm_hour == tm->tm_hour);
+            if (now - last_rt < 3000) ran_this_interval = true;
+        } else {
+            ran_this_interval = (ltm.tm_year == tm->tm_year && ltm.tm_yday == tm->tm_yday);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_scheduler.cpp:62]: Fixed. Removed the `now - last_rt < 3000` suppression for hourly schedules. Hourly now relies on interval identity (`year/yday/hour`) so next-hour runs are not incorrectly blocked.

File: src/clamwin-gui-cpp/resources/clamwin.manifest [hunk 16]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1"
+    xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
+    <assemblyIdentity
+        type="win32"
+        name="ClamWin.ClamWin"
+        version="0.103.2.0"
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/resources/clamwin.manifest]: Fixed. Manifest `assemblyIdentity` version is now aligned to `1.5.2.0` to match the GUI versioning used elsewhere.

File: src/clamwin-gui-cpp/cwdefs.h [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+#endif
+
+#ifndef WINVER
+#  define WINVER 0x0501
+#endif
+
+/* Require IE5.0+ headers for balloon tips, even on Win98 */
+#ifndef _WIN32_IE
+#  define _WIN32_IE 0x0500
+#endif
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cwdefs.h]: Fixed with Win98 compatibility preserved. `_WIN32_WINNT`/`WINVER` were restored to `0x0410` in `cwdefs.h`, and the comment now reflects Win98 baseline targeting.

File: cmake/legacy.cmake [hunk 18]
```diff
@@ -87,14 +87,27 @@ if(CLAMWIN_WINDOWS_VERSION LESS_EQUAL 0x0501 AND CLAMAV_ARCH STREQUAL "x86")
...
     get_target_property(RUST_ARCHIVE clamav_rust IMPORTED_LOCATION)
     set(RUST_FILTERED_ARCHIVE "${RUST_ARCHIVE}.filtered.a")
 
     add_custom_command(
         OUTPUT "${RUST_FILTERED_ARCHIVE}"
         COMMAND ${CMAKE_COMMAND} -E copy "${RUST_ARCHIVE}" "${RUST_FILTERED_ARCHIVE}"
-        COMMAND ${CMAKE_COMMAND} -E env python3 "${CMAKE_SOURCE_DIR}/filter-rust.py" "${CMAKE_OBJDUMP}" "${CMAKE_AR}" "${RUST_FILTERED_ARCHIVE}"
+        COMMAND ${CMAKE_COMMAND} -E env ${_RUST_FILTER_PY_CMD} "${CMAKE_SOURCE_DIR}/filter-rust.py" "${CMAKE_OBJDUMP}" "${CMAKE_AR}" "${RUST_FILTERED_ARCHIVE}"
         DEPENDS "${RUST_ARCHIVE}" "${CMAKE_SOURCE_DIR}/filter-rust.py"
     )
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply cmake/legacy.cmake:118]: Addressed. Python resolution is now explicit and robust (`find_program` for `py`/`python3`/`python`), and configuration now emits a clear `FATAL_ERROR` if no interpreter is available for `filter-rust.py`.

File: src/clamwin-gui-cpp/tests/test_update_checker.cpp [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+    CHECK(maj == 2);
+    CHECK(min == 5);
+    CHECK(pat == 0);
+}
+
+TEST_CASE("parseVersion: invalid input returns false")
+{
+    int maj = 0, min = 0, pat = 0;
+    CHECK_FALSE(CWUpdateChecker::parseVersion("", maj, min, pat));
+    CHECK_FALSE(CWUpdateChecker::parseVersion(NULL, maj, min, pat));
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/tests/test_update_checker.cpp]: Fixed. Replaced `NULL` with `nullptr` in the C++ test case for consistency and type safety.

File: src/clamwin-gui-cpp/cw_update_checker.cpp [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+
+#pragma comment(lib, "wininet.lib")
+
+/* ─── Hardcoded endpoints ───────────────────────────────────── */
+
+static const char* CW_GITHUB_API_HOST = "api.github.com";
+static const char* CW_GITHUB_API_PATH = "/repos/clamwin/clamwin/releases/latest";
+static const char* CW_DOWNLOAD_PAGE   = "https://www.clamwin.com/download";
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_update_checker.cpp:26]: Addressed. Added a legacy-OS gate before the GitHub HTTPS request path, so update checks are skipped on older systems lacking modern TLS support instead of attempting a failing API call.

File: src/clamwin-gui-cpp/cw_utils.cpp [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+                info->main_sigs = atoi(fields[3]);
+            }
+        }
+        CloseHandle(hFile);
+    }
+
+    /* Try daily.cld first, then daily.cvd */
+    wsprintfA(daily_path, "%s\\daily.cld", db_path);
+    if (GetFileAttributesA(daily_path) == INVALID_FILE_ATTRIBUTES)
+        wsprintfA(daily_path, "%s\\daily.cvd", db_path);
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_utils.cpp]: Fixed with the same bounded-path update in `CW_GetDBInfo`: path formatting now checks capacity and fails safely if composed DB paths would overflow.

File: src/clamwin-gui-cpp/cw_scheduler.cpp [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+        if (frequency == 3 && gap > 5400) time_to_run = true;
+        if (frequency == 0 && gap > 129600) time_to_run = true; /* 36h */
+        if (frequency == 1 && gap > 302400) time_to_run = true; /* 84h (3.5 days) */
+        if (frequency == 2 && gap > 648000) time_to_run = true; /* 7.5 days */
+    }
+
+    if (time_to_run) {
+        last_run_time = now;
+        return 1;
+    }
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_scheduler.cpp:122]: Fixed. The missed-run fallback now requires that the scheduled time has been reached before applying gap-based catch-up logic (`reached_scheduled_time` guard), preventing early pre-schedule runs.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+        ERROR_SUCCESS == RegOpenKeyEx(HKEY_LOCAL_MACHINE, shellExtRegPath, 0, KEY_READ, &hKey))
+    {
+        cbData = sizeof(szParams);
+        RegQueryValueEx(hKey, _T("params"), NULL, &dwType, (PBYTE)szParams, &cbData);
+        RegCloseKey(hKey);
+        // append params if exist
+        if (szParams[0] != _T('\0'))
+        {
+            // Expand Env
+            ExpandEnvironmentStrings(szParams, szParamsExpanded, sizeof(szParamsExpanded));
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp]: Fixed in the same change: both destination sizes now use `_countof(...)` (TCHAR count), avoiding Unicode-size mismatch with `ExpandEnvironmentStrings`.

File: src/clamwin-gui-cpp/cw_utils.cpp [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+    DWORD bytesRead;
+    char *p, *fields[8];
+    int i;
+
+    memset(info, 0, sizeof(*info));
+
+    /* Try main.cld first, then main.cvd */
+    wsprintfA(main_path, "%s\\main.cld", db_path);
+    if (GetFileAttributesA(main_path) == INVALID_FILE_ATTRIBUTES)
+        wsprintfA(main_path, "%s\\main.cvd", db_path);
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_utils.cpp]: Addressed. Replaced unbounded `wsprintfA` path composition with bounded `_snprintf` via a helper (`composeDbPath`), including truncation checks and graceful failure paths.

File: src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+        }
+    }
+    len = _tcslen(szClamWinPath);
+    if(!len)
+    {
+        MessageBox(hwnd, _T("Error: Unable to retrieve path to ClamWin. Please reinstall ClamWin Free Antivirus"), _T("ClamWin Free Antivirus"), MB_OK | MB_ICONERROR);
+        return FALSE;
+    }
+    // Expand Env
+    ExpandEnvironmentStrings(szClamWinPath, szPathExpanded, sizeof(szPathExpanded));
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/shell-extension/cw_shell_extension_impl.cpp]: Fixed. Updated both `ExpandEnvironmentStrings` calls to use TCHAR element counts via `_countof(...)` instead of `sizeof(...)` bytes:
- `szPathExpanded`: `_countof(szPathExpanded)`
- `szParamsExpanded`: `_countof(szParamsExpanded)`

File: src/clamwin-gui-cpp/cw_toggle.h [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+    /* Box / circle */
+    int markSize = CW_Scale(18);
+    RECT markRc;
+    markRc.left   = rc.left + CW_Scale(2);
+    markRc.top    = rc.top + (rc.bottom - rc.top - markSize) / 2;
+    markRc.right  = markRc.left + markSize;
+    markRc.bottom = markRc.top  + markSize;
+
+    COLORREF fill   = checked  ? theme->colorAccent()    : theme->colorSurface();
+    COLORREF border = disabled ? theme->colorTextMuted() : theme->colorTextMuted();
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_toggle.h]: Fixed. Simplified the dead conditional by giving distinct enabled/disabled border colors (`colorText()` vs `colorTextMuted()`), so the branch now has real effect and clearer intent.

File: src/clamwin-gui-cpp/cw_utils.cpp [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+        CloseHandle(hFile);
+    }
+
+    info->total_sigs = info->main_sigs + info->daily_sigs;
+    return (info->main_ver > 0 || info->daily_ver > 0) ? 1 : 0;
+}
+
+/* ─── Evaluate protection status ────────────────────────────── */
+
+CW_ProtectionStatus CW_GetProtectionStatus(const CWConfig *cfg)
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_utils.cpp:144]: Covered in the same test addition: `test_utils.cpp` exercises DB parsing/status paths and threshold transitions to reduce regression risk for this new logic.

File: src/clamwin-gui-cpp/cw_update_checker.cpp [hunk 28]
```diff
@@ -0,0 +1,278 @@
...
+
+    if (pos >= end || *pos != '"')
+        return false;
+    ++pos;  /* skip opening quote */
+
+    int i = 0;
+    while (pos < end && *pos != '"' && i < outCap - 1)
+    {
+        out[i++] = *pos++;
+    }
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_update_checker.cpp:170]: Fixed. `extractTagName()` now rejects truncated/unterminated values: it only succeeds when a closing quote is found before buffer capacity is exceeded.

File: src/clamwin-gui-cpp/cw_utils.cpp [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+                {
+                    FileTimeToSystemTime(&ft, &st);
+                    struct tm tm_val = {0};
+                    tm_val.tm_year = st.wYear - 1900;
+                    tm_val.tm_mon  = st.wMonth - 1;
+                    tm_val.tm_mday = st.wDay;
+                    tm_val.tm_hour = st.wHour;
+                    tm_val.tm_min  = st.wMinute;
+                    tm_val.tm_sec  = st.wSecond;
+                    info->updated_time = (long)mktime(&tm_val);
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_utils.cpp:131]: Addressed. Updated FILETIME conversion to local time before epoch calculation (`FileTimeToLocalFileTime` + `FileTimeToSystemTime`) so `mktime` no longer interprets UTC values as local and skews age calculations.

File: src/clamwin-gui-cpp/cw_utils.cpp [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+#include <time.h>
+
+/* ─── Parse a ClamAV .cvd/.cld header for version/sig info ─── */
+
+/*
+ * ClamAV CVD/CLD file header format (first 512 bytes):
+ *   ClamAV-VDB:build_time:version:num_sigs:...
+ * Fields separated by ':'
+ */
+int CW_GetDBInfo(const char *db_path, CW_DBInfo *info)
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_utils.cpp:39]: Added. I introduced `tests/test_utils.cpp` and wired it into `clamwin_gui_test` to cover `CW_GetDBInfo`/`CW_GetProtectionStatus` scenarios, including missing DB files and WARN/ERROR threshold behavior by DB age.

File: src/clamwin-gui-cpp/cw_time_edit.h [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+    char buf[32] = {0};
+    GetWindowTextA(edit, buf, sizeof(buf));
+    if (!buf[0]) return fallback;
+
+    bool isPM = false, isAM = false;
+    for (int i = 0; buf[i]; ++i)
+    {
+        if (buf[i] == 'P' || buf[i] == 'p') { isPM = true; break; }
+        if (buf[i] == 'A' || buf[i] == 'a') { isAM = true; break; }
+    }
```
[the remark this answers is not in the document]
Comment (alexcherney) [reply src/clamwin-gui-cpp/cw_time_edit.h]: Fixed. `cw_time_edit.h` now uses locale markers from `LOCALE_S1159` / `LOCALE_S2359` for both formatting and parsing, with case-insensitive matching, instead of hardcoded English AM/PM detection.

**Comment by sherpya** [comment]
@alexcherney please use this https://github.com/clamwin/clamwin, I'll try to see if it can be built for legacy later
